File: javascript/kontra/snake.js
// variables
SPRITE_SIZE = 20
SCREEN_SIZE = 20;
INITIAL_TAIL = 5;
FPS = 15

let { Sprite, GameLoop, initKeys } = kontra;
let { canvas, context } = kontra.init('surface');

let apple = Sprite({
    X: 3,
    Y: 3,
    width: SPRITE_SIZE - 1,
    height: SPRITE_SIZE - 1,
    color: 'red',

    render: function() {
        this.context.fillStyle = this.color;
        this.context.fillRect(this.X * SPRITE_SIZE, this.Y * SPRITE_SIZE, this.width, this.height);
    }
});

let snake = Sprite({
    X: 10,
    Y: 10,
    width: SPRITE_SIZE - 1,
    height: SPRITE_SIZE - 1,
    color: 'green',

    tail: INITIAL_TAIL,
    trail: [],

    checkCollision: function() {
        this.trail.forEach(element => {
            if(element.X == this.X && element.Y == this.Y) {
                if(snake.tail > INITIAL_TAIL){
                    console.log("Scored: %i", this.tail - INITIAL_TAIL);
                }
                this.X = this.Y = 10;
                this.velocity.x = this.velocity.y = 0;
                this.tail = INITIAL_TAIL;
            }
        });
    },

    update: function() {
        this.X = (this.X + this.velocity.x + SCREEN_SIZE) % SCREEN_SIZE;
        this.Y = (this.Y + this.velocity.y + SCREEN_SIZE) % SCREEN_SIZE;

        this.checkCollision();

        this.trail.push({X: this.X, Y: this.Y});
        while(this.trail.length > this.tail){
            this.trail.shift();
        }
    },

    render: function() {
        this.context.fillStyle = this.color;
        this.trail.forEach(element => {
            this.context.fillRect(element.X * SPRITE_SIZE, element.Y * SPRITE_SIZE, this.width, this.height);
        });
    }
});

function checkInput() {
    if(snake.velocity.x == 0){
        if(kontra.keyPressed('left')) {
            snake.velocity.x = -1;
            snake.velocity.y = 0;
        }
        if(kontra.keyPressed('right')) {
            snake.velocity.x = 1;
            snake.velocity.y = 0;
        }
    }
    if(snake.velocity.y == 0) {
        if(kontra.keyPressed('down')) {
            snake.velocity.x = 0;
            snake.velocity.y = 1;
        }
        if(kontra.keyPressed('up')) {
            snake.velocity.x = 0;
            snake.velocity.y = -1;
        }
    }
}

function generateApple() {
    needNewApple = false;
    do {
        needNewApple = false;
        apple.X = Math.floor(Math.random() * SCREEN_SIZE);
        apple.Y = Math.floor(Math.random() * SCREEN_SIZE);
        snake.trail.forEach(element => {
            needNewApple |= (element.X == apple.X && element.Y == apple.Y);
        });
    } while(needNewApple)
}

function checkPickup() {
    if(apple.X == snake.X && apple.Y == snake.Y){
        snake.tail++;
        generateApple();
    }
}

let loop = GameLoop({
    fps: FPS,
    clearCanvas: true,

    update: function() {
        checkInput();
        snake.update();
        checkPickup();
    },

    render: function() {
        apple.render();
        snake.render();
    }
});

initKeys();
loop.start();

File: golang/pixel/objects/snake.go
package objects

import (
	"image/color"
	"snake/constants"
	"snake/interfaces"
)

type Snake struct {
	position Vector
	velocity Vector
	width    int
	height   int
	color    color.RGBA
	tail     int
	trail    []Vector
}

func (s *Snake) CheckCollision(position Vector) bool {
	for _, p := range s.trail {
		if p.X == position.X && p.Y == position.Y {
			return true
		}
	}
	return false
}

func (s *Snake) handleInput(inputHandler interfaces.InputHandler) {
	directionPressed := inputHandler.HandleInput()
	if directionPressed == interfaces.Left && s.velocity.X == 0 {
		s.velocity.X = -1
		s.velocity.Y = 0
	}
	if directionPressed == interfaces.Right && s.velocity.X == 0 {
		s.velocity.X = 1
		s.velocity.Y = 0
	}
	if directionPressed == interfaces.Down && s.velocity.Y == 0 {
		s.velocity.X = 0
		s.velocity.Y = 1
	}
	if directionPressed == interfaces.Up && s.velocity.Y == 0 {
		s.velocity.X = 0
		s.velocity.Y = -1
	}
}

func (s *Snake) Draw(renderer interfaces.Renderer) {
	for _, p := range s.trail {
		x := p.X * constants.SpriteSize
		y := p.Y * constants.SpriteSize
		renderer.RenderRectangle(x, y, x+s.width, y+s.height, s.color)
	}
}

func (s *Snake) move() {
	s.position.X = (s.position.X + s.velocity.X + constants.ScreenSize) % constants.ScreenSize
	s.position.Y = (s.position.Y + s.velocity.Y + constants.ScreenSize) % constants.ScreenSize
}

func (s *Snake) reset() {
	s.position = Vector{constants.ScreenSize / 2, constants.ScreenSize / 2}
	s.velocity = Vector{0, 0}
	s.tail = constants.InitialTail
}

func (s *Snake) trimTrail() {
	s.trail = append(s.trail, s.position)
	if len(s.trail) > s.tail {
		ret := make([]Vector, 0)
		s.trail = append(ret, s.trail[1:]...)
	}
}

func (s *Snake) Update(inputHandler interfaces.InputHandler) {
	s.handleInput(inputHandler)
	s.move()
	if s.CheckCollision(s.position) {
		s.reset()
	}
	s.trimTrail()
}
